api/src/main/java/org/hawkular/inventory/api/filters/Filter.java

```diff
@@ -42,9 +42,10 @@
 public abstract class Filter {
     private static final Filter[] EMPTY = new Filter[0];
 
-    Filter() {
-
-    }
+//    we should not prohibit creating sub-classes of Filter in the implementations of this API
```

Comment:
True, but on the other hand, this enables the user of the library to do the same and expect the implementation would work with it.

We could make that work by adding `boolean apply(Entity)` to the `Filter` class and implement an "in-memory" application of the filters. Implementations of the API can then choose to implement the specific filter classes in a more efficient manner and fall back to the `apply` method otherwise.

But I am not sure if that's really advisable solution. Our primary concern is the REST API which would lack such possibility.

Not sure how to approach this though. Your need for custom filters clear but at the same time I would not like to open the possibility for user-defined Filter impls. Maybe the `PathContext` and the accompanied classes would not use the `Filter`s directly but some kind of Filter decorator that would allow using the predefined and final set of filters from the API yet would allow for custom filters defined inside the implementation non-publicly? But that, too, is quite a big change.